File: Leetcode/Q_23_Merge_k_Sorted_Lists/Sample_test.cpp
#include "Sample.h"

#include "gmock/gmock.h"
#include <gtest/gtest.h>
#include<iostream>
#include<list>

using ::sample::Solution;
using namespace sample;

ListNode* makeListFromVector(std::vector<int> v) {
    ListNode* last = nullptr;
    for (std::vector<int>::reverse_iterator it = v.rbegin(); it != v.rend(); it++) {
        last = new ListNode(*it, last);
    }
    return last;
}

void DeleteList(ListNode* head) {
    std::list<ListNode*> nodes;
    ListNode* current = head;
    while (current != nullptr) {
        nodes.push_back(current);
        current = current->next;
    }
    for (std::list<ListNode*>::reverse_iterator it = nodes.rbegin(); it != nodes.rend(); it++) {
        delete *it;
    }
}

std::vector<int> makeVectorFromList(ListNode* head) {
    ListNode* current = head;
    std::vector<int> v;
    while (current != nullptr) {
        v.push_back(current->val);
        current = current->next;
    }
    return v;
}

TEST(Sample, Example1) {
    std::vector<ListNode*> input = {
        makeListFromVector({1,4,5}),
        makeListFromVector({1,3,4}),
        makeListFromVector({2,6})
    };
    ListNode* result = Solution().mergeKLists(input);
    EXPECT_THAT(makeVectorFromList(result), ::testing::ElementsAre(1,1,2,3,4,4,5,6));
    DeleteList(result);
}

TEST(Sample, Example2) {
    std::vector<ListNode*> input = {};
    ListNode* result = Solution().mergeKLists(input);
    std::vector<int> result_vector = makeVectorFromList(result);
    EXPECT_THAT(std::vector<int>(), ::testing::IsEmpty());
    DeleteList(result);
}

TEST(Sample, Example3) {
    std::vector<ListNode*> input = {
        makeListFromVector({})
    };
    ListNode* result = Solution().mergeKLists(input);
    EXPECT_THAT(makeVectorFromList(result), ::testing::IsEmpty());
    DeleteList(result);
}

File: Leetcode/Q_25_Reverse_Nodes_in_k-Group/sample_test.py
import unittest
import sample
from sample import ListNode

class TestReverseKGroup(unittest.TestCase):
    '''Unit test for sample reverseKGroup function.'''

    def get_list_node_vals(self, head: ListNode):
        current = head
        result = []
        while current != None:
            result.append(current.val)
            current = current.next
        return result
    
    def test_example_1(self):
        n1 = ListNode(val = 1)
        n2 = ListNode(val = 2)
        n3 = ListNode(val = 3)
        n4 = ListNode(val = 4)
        n5 = ListNode(val = 5)
        n1.next = n2
        n2.next = n3
        n3.next = n4
        n4.next = n5
        expected = [2, 1, 4, 3, 5]
        actual = self.get_list_node_vals(sample.Solution().reverseKGroup(n1, 2))
        self.assertEqual(actual, expected)

    def test_example_2(self):
        n1 = ListNode(val = 1)
        n2 = ListNode(val = 2)
        n3 = ListNode(val = 3)
        n4 = ListNode(val = 4)
        n5 = ListNode(val = 5)
        n1.next = n2
        n2.next = n3
        n3.next = n4
        n4.next = n5
        expected = [3, 2, 1, 4, 5]
        actual = self.get_list_node_vals(sample.Solution().reverseKGroup(n1, 3))
        self.assertEqual(actual, expected)

if __name__ == "__main__":
    unittest.main()
